import_obfuscation.go

```diff
@@ -0,0 +1,491 @@
+			// import path
+			sort.Slice(privateImports, func(i, j int) bool {
+				iSlashes := strings.Count(privateImports[i], "/")
+				jSlashes := strings.Count(privateImports[j], "/")
+				// sort by number of slashes first, then alphabetically
+				if iSlashes == jSlashes {
+					return privateImports[i] > privateImports[j]
+				}
+				return iSlashes > jSlashes
+			})
+			privateImports = dedupImportPaths(privateImports)
+
+			// no private import paths, nothing to garble
+			// log.Printf("== Private imports: %v ==\n", privateImports)
+			if len(privateImports) == 0 {
+				continue
+			}
+
+			// garble all private import paths in all symbol names
+			lists := [][]*goobj2.Sym{am.SymDefs, am.NonPkgSymDefs, am.NonPkgSymRefs}
+			for _, list := range lists {
+				for _, s := range list {
+					// garble read only static data, but not strings. If import paths are in strings,
+					// that means garbling strings might effect the behavior of the compiled binary
+					if s.Kind == 2 && s.Data != nil && !strings.HasPrefix(s.Name, "go.string.") {
+						var dataTyp dataType
+						if strings.HasPrefix(s.Name, "type..importpath.") {
+							dataTyp = importPath
+						} else if strings.HasPrefix(s.Name, "type..namedata.") {
+							dataTyp = namedata
+						}
+						s.Data = garbleSymData(s.Data, privateImports, garbledImports, dataTyp, &buf)
+
+						if s.Size != 0 {
+							s.Size = uint32(len(s.Data))
+						}
+					}
+					s.Name = garbleSymbolName(s.Name, privateImports, garbledImports, &sb)
+
+					for i := range s.Reloc {
+						s.Reloc[i].Name = garbleSymbolName(s.Reloc[i].Name, privateImports, garbledImports, &sb)
+					}
+					if s.Type != nil {
+						s.Type.Name = garbleSymbolName(s.Type.Name, privateImports, garbledImports, &sb)
+					}
+					if s.Func != nil {
+						for i := range s.Func.FuncData {
+							s.Func.FuncData[i].Sym.Name = garbleSymbolName(s.Func.FuncData[i].Sym.Name, privateImports, garbledImports, &sb)
+						}
+						for _, inl := range s.Func.InlTree {
+							inl.Func.Name = garbleSymbolName(inl.Func.Name, privateImports, garbledImports, &sb)
+						}
+					}
+				}
+			}
+			for i := range am.SymRefs {
+				am.SymRefs[i].Name = garbleSymbolName(am.SymRefs[i].Name, privateImports, garbledImports, &sb)
+			}
+
+			if err := p.pkg.Write(p.path); err != nil {
+				return nil, fmt.Errorf("error writing objfile %s at %s: %v", p.pkg.ImportPath, p.path, err)
+			}
+		}
+	}
+
+	// garble importcfg so the linker knows where to find garbled imports
+	for pkgPath, info := range importCfg {
+		if isPrivate(pkgPath) {
+			pkgPath = hashImport(pkgPath, garbledImports)
+		}
+		if info.IsSharedLib {
+			buf.WriteString("packageshlib")
+		} else {
+			buf.WriteString("packagefile")
+		}
+
+		buf.WriteRune(' ')
+		buf.WriteString(pkgPath)
+		buf.WriteRune('=')
+		buf.WriteString(info.Path)
+		buf.WriteRune('\n')
+	}
+
+	if err := ioutil.WriteFile(importCfgPath, buf.Bytes(), 0644); err != nil {
```

Comment:
FYI, it would be better to use a bufio.Writer directly on top of os.Create; that way you don't need to hold the entire buffer in memory at once.